build_tools/cmake/iree_macros.cmake

```diff
@@ -62,11 +62,15 @@ function(iree_package_ns PACKAGE_NS)
   string(REPLACE ${IREE_ROOT_DIR} "" _RELATIVE_PATH ${CMAKE_CURRENT_LIST_DIR})
   string(SUBSTRING ${_RELATIVE_PATH} 1 -1 _RELATIVE_PATH)
 
+  if(NOT "${IREE_ROOT_DIR}/build_tools/cmake" STREQUAL ${CMAKE_CURRENT_FUNCTION_LIST_DIR})
+    # If this macro file is not under IREE_ROOT_DIR, the function is used
+    # out-of-tree and the relative path should be used directly.
+    set(_PACKAGE "${_RELATIVE_PATH}")
```

Comment:
I'd like to iterate on this myself, but I don't have a repo set up for it right now...

I'd recommend checking something like
```cmake
if(NOT ${CMAKE_CURRENT_LIST_DIR} MATCHES "^${IREE_ROOT_DIR}/.*")
```

I'm not sure about using `CMAKE_CURRENT_FUNCTION_LIST_DIR` or anything specifically referencing out of tree usage in that way.

Reply:
To use `iree_macros.cmake` correctly, in-project or out-of-tree, `IREE_ROOT_DIR` needs to be set to top-level CMakeLists project, i.e., `CMAKE_SOURCE_DIR`. (Actually, the reason is at line 62 of this snippet). In that case, `CMAKE_CURRENT_LIST_DIR` should always start with `IREE_ROOT_DIR`

`CMAKE_CURRENT_FUNCTION_LIST_DIR` is the path of where this iree_macros.cmake file sits, regardless where the function is called. I can change the check `CMAKE_CURRENT_FUNCTION_LIST_DIR` and `IREE_ROOT_DIR`, but without the specific path the condition would not work if IREE is included as a submodule and is placed under `<top_project>/third_party/iree`

Reply:
It is possible to update the hard-coded usage of `IREE_ROOT_DIR` in Line 62 so we don't need to apply this hack, then what you propose would work. I can update this PR to fix both.

Reply:
Ah. How about using the `CMAKE_SOURCE_DIR`-relative path if not under `IREE_ROOT_DIR`?

Reply:
maybe with two variables and some descriptive names?

```cmake
function(iree_package_ns PACKAGE_NS)
  # Get the IREE-relative path of the current dir (i.e. runtime/src/iree/vm).
  string(REPLACE ${IREE_ROOT_DIR} "" _IREE_RELATIVE_PATH ${CMAKE_CURRENT_LIST_DIR})
  string(SUBSTRING ${_IREE_RELATIVE_PATH } 1 -1 _IREE_RELATIVE_PATH)

  if(NOT ${CMAKE_CURRENT_LIST_DIR} MATCHES "^${IREE_ROOT_DIR}/.*")
    # Function is being called from outside IREE. Use the source-relative path.
    string(REPLACE ${CMAKE_SOURCE_DIR} "" _SOURCE_RELATIVE_PATH ${CMAKE_CURRENT_LIST_DIR})
    string(SUBSTRING ${_SOURCE_RELATIVE_PATH} 1 -1 _SOURCE_RELATIVE_PATH)
    set(_PACKAGE "${_SOURCE_RELATIVE_PATH}")
  elseif(_IREE_RELATIVE_PATH MATCHES "^compiler/src/(.*)")
    ...
```